Extend the action manager to work with layer.

Carreau opened the pull request:
This is mostly in flight thoughts, nothing concrete yet,but happy to get
feedback on the overall approach.

 - This does not yet remove the usage of KeymapProvider but will likely
   will have to be added at some point.
 - Action Manager Context is now a mapping like, except that if value
   are callable they are going to be evaluated before being passed to
   action. In particular we use this with `layer` to make it pass the
   currently active layer. I guess they could take viewer and manipulate
   viewer.selection.layers directly...I'm wondering if a more general
   notion of classes with multiple instance and which one are
   focused/active/selected would be useful.
 - I haven't touch non-Label layer, technically a action that ask for
   'layer' could be passed a layer type it does not know about, that's
   where Kira comment on previous PR about using type annotations woudl
   be useful.
 - We still do not deal with function generator that trigger an action
   on press and release. For now I've added an option to add extra text
   when binding a button to keep the "toggle with {shortcut}".
 - I also inject action_manager in the qtconsole context, so one does
   not have to import it everytime,
 - there is currently a _debug(bool) method on the action manager the
   will update all the tooltips to insert the name of the action, that's
   mostly to debug the UI and will be removed.
 - I'm not sure how we want to deal with spinlabels and have +/- showing
   tooltips, I discoverd the I/D would increase/decrease the label values,
   but it is nowhere in the UI ?
 - the bind_button seem to work with QtModeRadioButton, but that seem to
   be a fluke as it is not meant to, so I need to dive into this before.
 - Layer have undo/redo bound to Ctrl-Z/Ctrl-Shift-Z; I find strange
   that this is not at the global level , that seem like a codesmell so
   I've left it alone for now.

# Description
<!-- What does this pull request (PR) do? Why is it necessary? -->
<!-- Tell us about your new feature, improvement, or fix! -->
<!-- If your change includes user interface changes, please add an image, or an animation "An image is worth a thousando words!" -->
<!-- You can use https://www.cockos.com/licecap/ or similar to create animations -->

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

goanpeca commented on napari/_qt/layer_controls/qt_labels_controls.py:
```diff
@@ -160,24 +162,37 @@ def __init__(self, layer):
                 shortcut=Shortcut("Control"),
             ),
         )
+        # don't bind with action manager as this would remove "Toggle with {shortcut}"
```
I don't understand the comment @Carreau 🤔, could you explain :)

Carreau replied:
Remainders of earlier commit, removed.

goanpeca commented on napari/_qt/layer_controls/qt_labels_controls.py:
```diff
@@ -160,24 +162,37 @@ def __init__(self, layer):
+
+        self.pick_button = QtModeRadioButton(layer, 'picker', Mode.PICK)
+        action_manager.bind_button(
+            'napari:activate_picker_mode', self.pick_button
```
In general I would prefer we had all these strings defined as "enums", as it would make it easier to change eventually (if it made sense) and also avoid typing mistakes?

```python
class NapariLayerActions:

     ActivatePaintMode = 'napari:activate_paint_mode'
     ActivatePickerMode = 'napari:activate_picker_mode'


```

Carreau replied:
I'm not sure this is worth it; 1) the registration is based on the function name, and the decorator automatically prepend `"napari:"`, so you would still need to make sure the values in the enum are the same as the function name 2) you won't be able to use the enum in the settings anyway. 3) That would defeat the purpose of bing able to define/bind in any order as now you need both the defining location and binding location to be able to import the enum.

goanpeca reviewed (approved):
LGTM @Carreau, thanks 🚀

sofroniewn commented on napari/layers/labels/_labels_key_bindings.py:
```diff
@@ -16,51 +19,79 @@ def hold_to_pan_zoom(layer):
 
 
-@Labels.bind_key('P')
+def register_label_action(description, shortcuts):
```
Can we turn these into a single `register_layer_action(layer_type, description, shortcuts)`. That seems like it will be much more reusable across the code base (could go in `layer_utils.py` file)

Carreau replied:
I can, though I'll expect the end API to change a bit with the action manager handling the keybinding.

sofroniewn replied:
Yeah, i think the current approach is ok. You could also make all these private `_register_label_action` for example, I don't have super strong feelings though, curious if others do

sofroniewn commented on napari/layers/labels/_labels_key_bindings.py:
```diff
@@ -1,3 +1,6 @@
+from napari.utils.action_manager import action_manager
```
We use relative imports in files like this

sofroniewn commented on napari/_qt/widgets/qt_viewer_buttons.py:
```diff
@@ -102,6 +102,14 @@ def __init__(self, viewer):
         self.viewer = viewer
         action_manager.context['viewer'] = viewer
 
+        def active_layer():
```
should this take `self`, and use `self.viewer` everywhere?

Carreau replied:
Sorry for the delay in responding, some issues with travelling and just starting to get back online today.

`action_manager.context` can either takes objects or callables with no arguments. So this is a closure.
I could change how action_manager's context works.

Carreau replied:
(but yet to self.viewer everywhere)

sofroniewn replied:
i'm ok with this, just wanted to check! we can leave open in case others have thoughts here too

sofroniewn reviewed (commented):
Looking good! I think it's pretty close, my big request would be if we could get away without all the `register_label_action`, `register_points_action`, ... and just have one for `register_layer_action` that takes a layer type.

sofroniewn reviewed (changes requested):
See comments above, in particular creating single `register_layer_action` function

Carreau committed:
Extend the action manager to work with layer.

This is mostly in flight thoughts, nothing concrete yet,but happy to get
feedback on the overall approach.

 - This does not yet remove the usage of KeymapProvider but will likely
   will have to be added at some point.
 - Action Manager Context is now a mapping like, except that if value
   are callable they are going to be evaluated before being passed to
   action. In particular we use this with `layer` to make it pass the
   currently active layer. I guess they could take viewer and manipulate
   viewer.selection.layers directly...I'm wondering if a more general
   notion of classes with multiple instance and which one are
   focused/active/selected would be useful.
 - I haven't touch non-Label layer, technically a action that ask for
   'layer' could be passed a layer type it does not know about, that's
   where Kira comment on previous PR about using type annotations woudl
   be useful.
 - We still do not deal with function generator that trigger an action
   on press and release. For now I've added an option to add extra text
   when binding a button to keep the "toggle with {shortcut}".
 - I also inject action_manager in the qtconsole context, so one does
   not have to import it everytime,
 - there is currently a _debug(bool) method on the action manager the
   will update all the tooltips to insert the name of the action, that's
   mostly to debug the UI and will be removed.
 - I'm not sure how we want to deal with spinlabels and have +/- showing
   tooltips, I discoverd the I/D would increase/decrease the label values,
   but it is nowhere in the UI ?
 - the bind_button seem to work with QtModeRadioButton, but that seem to
   be a fluke as it is not meant to, so I need to dive into this before.
 - Layer have undo/redo bound to Ctrl-Z/Ctrl-Shift-Z; I find strange
   that this is not at the global level , that seem like a codesmell so
   I've left it alone for now.

register action for most of the shape layer

bug

rename for test

rename in test

Misc Cleanup

Start to extends to points layer

New example

update docstrings and binding

empty

take review into accounts

Carreau committed:
flake8

Carreau committed:
black

Carreau committed:
isort

sofroniewn reviewed (approved):
Ok, this looks good to me! My comments now are minor, we can continue to improve as we go. Maybe either @jni and @tlambert03 can review too and then we can merge!!
